pluggable-triggers.md

```diff
@@ -0,0 +1,205 @@
+   private List<SObject> records = new List<SObject>();
+
+
+   public TriggerName(List<SObject> records){
+       this.records = records;
+   }
+  
+   public void onBeforeInsert() {
+      
+   }
+
+
+   public void onBeforeUpdate(Map<Id, SObject> existingRecords) {
+      
+   }
+
+
+   public void onBeforeDelete() {
+      
+   }
+
+
+   public void onAfterInsert() {
+      
+   }
+
+
+   public void onAfterUpdate(Map<Id, SObject> existingRecords) {
+      
+   }
+
+
+   public void onAfterDelete() {
+      
+   }
+
+
+   public void onAfterUndelete() {
+      
+   }
+}
+````
+
+2. Create the trigger constructor.
+
+The trigger constructor is used to create instances of your trigger plugin from step one. See the API documentation for fferpcore.PluggableTriggerApi.PluginConstructor for more details. An example implementation is provided below - the reference to fferpcore__BillingDocument__c should be changed to your target SObject. Note that this class and the methods within it must be declared as global so that the Foundations package can invoke them.
+
+````
+global class TriggerNameConstructor
+   implements fferpcore.PluggableTriggerApi.PluginConstructor {
+
+
+   global SObjectType sObjectType() {
+       // The SObjectType for the SObject you want to add the trigger to,
+       // e.g. Billing Document.
+       return fferpcore__BillingDocument__c.SObjectType;
+   }
+
+
+   global fferpcore.PluggableTriggerApi.Plugin construct(
+       List<SObject> records,
+       fferpcore.PluggableTriggerApi.Context context // Currently unused
+   ) {
+       return new TriggerName(records);
+   }
+}
+````
+
+3. Add an Apex Trigger for the target SObject (can be skipped if the SObject already uses the pluggable trigger framework).
+
+This trigger should invoke the construct() method on the fferpcore.PluggableTrigger.Constructor() class, which will find all the corresponding trigger plugins and execute them in order. The methods on the fferpcore.PluggableTrigger instance this returns should be invoked accordingly. We recommend covering all operation types even if they are not required yet to avoid confusion in the future. You do not need to add an Apex Trigger if one already exists on this SObject, either in a Certinia managed package or unmanaged on your org. See the list of existing Certinia pluggable triggers. 
+
+````
+trigger BillingDocumentTrigger on fferpcore__BillingDocument__c(
+   before insert,
+   before update,
+   before delete,
+   after insert,
+   after update,
+   after delete,
+   after undelete
+) {
+   List<SObject> records;
+
+
+   switch on Trigger.operationType {
+       when BEFORE_DELETE, AFTER_DELETE {
+           records = Trigger.old;
+       }
+       when else {
+           records = Trigger.new;
+       }
+   }
+
+
+   try {
+       fferpcore.PluggableTrigger target = 
+              new fferpcore.PluggableTrigger.Constructor()
+                     .construct(records);
+
+
+       switch on Trigger.operationType {
```

Comment:
A recent code change in `fferpcore` added a global method `runTriggerHandler()`. This is slightly different in that it delegates back to `fflib_SobjectDomain` to instantiate the `PluggableTrigger` then follows the SObjectDomain patterns for calling methods. The methods it calls are slightly different to here, though the differences may be that this code does not call methods that `PluggableTrigger` has not provided implementation for.

I don't know if we want to document the new `runTriggerHandler()` method instead? It would reduce boilerplate code.

Reply:
That's a good point. I think with the new runTriggerHandler() method we can greatly simplify the Apex trigger code snippet provided here. Are you happy if I update this on/after Summer '24 ERD?